### PR TITLE
k_heap: Clamp to a minimum heap size (and add test)

### LINKED ISSUE #33009: kernel: k_heap failures on small heaps
**Describe the bug**

`k_heap_alloc` fails when attempting to allocate memory from small heaps.
The following code will fail:
```
K_HEAP_DEFINE(test_heap, 40);

void main(void) {
    void * p = k_heap_alloc(&test_heap, 8, K_NO_WAIT);
    printk("%p\n", p);
}
```
On a Cortex M4F board, this will just return NULL, but on `native_posix_64` a `K_ERR_KERNEL_PANIC` is triggered.
Based on the documentation (https://docs.zephyrproject.org/latest/reference/kernel/memory/heap.html), it seems clear that there is simply not enough space at the start of the buffer to hold the bucket list for allocations.

The bug is not that the k_heap cannot support tiny heaps, but that fatal errors can occur.
It also not obvious from the documentation how large the heap needs to be before it will start working.
64 bytes is sufficient to allocate 8 bytes on a Cortex M4F , but not `native_posix_64`.

**Expected behavior**

Allocating from small heap should just return NULL, not hard fault.
Ideally some form of validation would be done on the values provided to `K_HEAP_DEFINE` to ensure that it is large enough to return a valid 8 byte (smallest allocation unit) memory chunk at least once.

Something similar to:
```
#define K_HEAP_DEFINE(name, bytes)				\
	BUILD_ASSERT((bytes) >= (MINIMUM_BUCKET_SIZE + 8), "Heap too small to allocate memory from");\
	char __aligned(sizeof(void *)) kheap_##name[bytes];	\
	Z_STRUCT_SECTION_ITERABLE(k_heap, name) = {		\
		.heap = {					\
			.init_mem = kheap_##name,		\
			.init_bytes = (bytes),			\
		 },						\
	}
```

**Environment (please complete the following information):**
 - Zephyr v2.5

### PR DESCRIPTION
**Pull request by andyross**
k_heap: Clamp to a minimum heap size

The K_HEAP_DEFINE macro would allow users to specify heaps that are
too small, leading to potential corruption events (though at least
there were __ASSERTs that would catch this at runtime if enabled).

It would be nice to put the logic to compute this value into the heap
code, but that isn't available in kernel.h (and we don't want to pull
it in as this header is already WAY to thick).  So instead we just
hand-compute and document the choice.  We can address bitrot problems
with a test.

Fixes #33009

### CODE REVIEW

**Comment by andyross** [comment]
In point of fact, the minimum sizes here are the result of hand-probing for the boundary, and somewhat larger than what I hand-calculated.  It seems in pathological situations like this we're creating more bucket levels than needed (if you only have one usable chunk, you should only need one level).  I didn't want to try to fix it lest I manage to forget some edge case right before release and break something, the value of the fix just isn't that high.  But maybe worth investigating later.

File: tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c [hunk 1]
```diff
@@ -40,6 +40,40 @@ static void thread_alloc_heap(void *p1, void *p2, void *p3)
 
 /*test cases*/
 
+/* These need to be adjacent in BSS */
+volatile uint32_t heap_guard0;
+K_HEAP_DEFINE(tiny_heap, 1);
+volatile uint32_t heap_guard1;
```
Comment (dcpleung) [inline tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c:46]: Hm... can toolchain relocate them freely, so these won't be together?

Comment (andyross) [reply tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c:46]: They will on existing platforms.  You're right that with -fdata-sections you could write a linker script to break this.  But in practice linker directives that select symbols with the same wildcard preserve their order within the file, and it's no doubt to preserve behavior in code exactly like this.

And in any case, the side effect would be that the weird platform with a pathological linker script gets a false "pass" result on this particular test case in the face of an overrun that gets caught on more conventional platforms.

I'm comfortable with it, anyway :)

**Review by dcpleung — APPROVED** [review]

**Comment by npitre** [comment]
What about making a second 1-byte allocation and verify that it does indeed
fail? This way you'll make sure that the minimum size is indeed the minimum.

**Comment by andyross** [comment]
To be clear: those values (56/64 for big/small heaps) were chosen because 55/63 were experimentally verified as failing setup due to the existing size asserts (which certainly look correct).  But by my hand-calculation, it ought to actually be 48 bytes for a big heap.

I **think** the difference is that the number of buckets calculation is done on the basis of the full heap size and not the usable heap size, because the latter has to be defined recursively because more buckets means less heap.  And that means we end up with too many buckets in the case of pathologically small heaps.

Now... is 16 bytes on a heap that isn't actually useful for anything worth fretting over?   Probably not, and absolutely not right now before a release.  Was just an interesting thing I noticed. :)

**Comment by galak** [comment]
@JordanYates fyi.

**Comment by JordanYates** [comment]
Looks like a reasonable solution to me

**Review by nashif — APPROVED** [review]

**Comment by npitre** [comment]
OK... I don't like it when something is weird without a good explanation.
So I took upon myself to find what is going on here. In doing so I noted
two problems with the heap code that your test made visible. Those are
very minor issues but still.

Firstly:

```
diff --git a/include/kernel.h b/include/kernel.h
index e5c2e44a39..ee00bd71a1 100644
--- a/include/kernel.h
+++ b/include/kernel.h
@@ -5009,7 +5009,7 @@ void k_heap_free(struct k_heap *h, void *mem);
  * @param bytes Size of memory region, in bytes
  */
 #define K_HEAP_DEFINE(name, bytes)				\
-	char __aligned(sizeof(void *))				\
+	char __aligned(8) /* CHUNK_UNIT */			\
 	     kheap_##name[MAX(bytes, Z_HEAP_MIN_SIZE)];		\
 	Z_STRUCT_SECTION_ITERABLE(k_heap, name) = {		\
 		.heap = {					\
```

The heap memory is internally aligned to CHUNK_UNIT and any external
alignment smaller than that is useless. At best it will do nothing, at
worst it will waste some memory needlessly. The linker normally sorts
objects by their alignment so a bigger alignment on 32-bit system would
always be optimal.

And having a random 4-bytes alignment does make your test unstable, as
sometimes the minimal size is OK, sometimes it is not, depending on the
arcchitecture which makes it even more confusing. I know as I've been
bitten by that kind of varying heap placement before.

Secondly:

```
diff --git a/lib/os/heap.c b/lib/os/heap.c
index 2dee196404..ba1e9629c4 100644
--- a/lib/os/heap.c
+++ b/lib/os/heap.c
@@ -403,7 +403,7 @@ void sys_heap_init(struct sys_heap *heap, void *mem, size_t bytes)
 	chunksz_t chunk0_size = chunksz(sizeof(struct z_heap) +
 				     nb_buckets * sizeof(struct z_heap_bucket));
 
-	__ASSERT(chunk0_size + min_chunk_size(h) < heap_sz, "heap size is too small");
+	__ASSERT(chunk0_size + min_chunk_size(h) <= heap_sz, "heap size is too small");
 
 	for (int i = 0; i < nb_buckets; i++) {
 		h->buckets[i].next = 0;
```

Strictly speaking, min_chunk_size() is... well... the minimum chunk size that
can be allocated. So it is OK to be equal to the heap size here.

With those 2 fixes in place, the formal minimum size values become:

```
diff --git a/include/kernel.h b/include/kernel.h
index e5c2e44a39..067d49228c 100644
--- a/include/kernel.h
+++ b/include/kernel.h
@@ -4992,7 +4992,7 @@ void k_heap_free(struct k_heap *h, void *mem);
 /* Hand-calculated minimum heap sizes needed to return a successful
  * 1-byte allocation.  See details in lib/os/heap.[ch]
  */
-#define Z_HEAP_MIN_SIZE (sizeof(void*) > 4 ? 64 : 56)
+#define Z_HEAP_MIN_SIZE (sizeof(void*) > 4 ? 56 : 44)
 
 /**
  * @brief Define a static k_heap
```

And then you may add this to your test to fully validate the above:

```
--- a/tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c
+++ b/tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c
@@ -64,8 +64,10 @@ void test_k_heap_min_size(void)
 	heap_guard1 = guard_bits;
 
 	char *p = k_heap_alloc(&tiny_heap, 1, K_FOREVER);
+	char *q = k_heap_alloc(&tiny_heap, 1, K_NO_WAIT);
 
 	zassert_not_null(p, "allocation failed");
+	zassert_is_null(q, "allocation did not fail as expected");
 
 	*p = 0xff;
 	k_heap_free(&tiny_heap, p);
```

By the way, it is probably a good idea to change that `K_FOREVER` above
to `K_NO_WAIT`, otherwise the assertion on `p` will never have a chance
to fail.

**Comment by andyross** [comment]
Makes sense to me.  Pulled in @npitre's changes more or less unmodified.

**Review by laurenmurphyx64 — APPROVED** [review]

**Comment by nashif** [comment]
@npitre  you are good?

**Review by npitre — APPROVED** [review]
Yes, lgtm
